docs/swift/DesignDoc.md

```diff
@@ -0,0 +1,674 @@
+struct, subscript, super, switch, throw, throws, true, try, typealias, var,
+where, while
+```
+
+The set of contextually reserved words is:
+
+```
+associativity, convenience, dynamic, didSet, final, get, infix, indirect,
+lazy, left, mutating, none, nonmutating, optional, override, postfix,
+precedence, prefix, Protocol, required, right, set, Type, unowned, weak,
+willSet
+```
+
+It is possible to use any reserved word as an identifier by escaping it with
+backticks (for example, ``let `class` = 5``). Other name-mangling schemes would
+require us to transform the names themselves (for example, by appending an
+underscore), which requires us to then ensure that the new name does not collide
+with something else in the same namespace.
+
+While the backtick feature may not be widely known by all Swift developers, a
+small amount of user education can address this and it seems like the best
+approach. We can unconditionally surround all property names with backticks to
+simplify generation.
+
+Some remapping will still be required, though, to avoid collisions between
+generated properties and the names of methods and properties defined in the base
+protocol/implementation of messages.
+
+# Features of Protocol Buffers
+
+This section describes how the features of the protocol buffer syntaxes (proto2
+and proto3) map to features in Swift—what the code generated from a proto will
+look like, and how it will be implemented in the underlying library.
+
+## Packages
+
+Modules are the main form of namespacing in Swift, but they are not declared
+using syntactic constructs like namespaces in C++ or packages in Java. Instead,
+they are tied to build targets in Xcode (or, in the future with open-source
+Swift, declarations in a Swift Package Manager manifest). They also do not
+easily support nesting submodules (Clang module maps support this, but pure
+Swift does not yet provide a way to define submodules).
+
+We will generate types with fully-qualified underscore-delimited names. For
+example, a message `Baz` in package `foo.bar` would generate a struct named
+`Foo_Bar_Baz`. For each fully-qualified proto message, there will be exactly one
+unique type symbol emitted in the generated binary.
+
+Users are likely to balk at the ugliness of underscore-delimited names for every
+generated type. To improve upon this situation, we will add a new string file
+level option, `swift_package_typealias`, that can be added to `.proto` files.
+When present, this will cause `typealias`es to be added to the generated Swift
+messages that replace the package name prefix with the provided string. For
+example, the following `.proto` file:
+
+```protobuf
+option swift_package_typealias = "FBP";
+package foo.bar;
+
+message Baz {
+  // Message fields
+}
+```
+
+would generate the following Swift source:
+
+```swift
+public struct Foo_Bar_Baz {
+  // Message fields and other methods
+}
+
+typealias FBPBaz = Foo_Bar_Baz
```

Comment:
In most deserialization contexts it doesn't matter because generated code knows the type of nested messages to parse, even in text/JSON format scenarios, but the assumption of uniqueness of a message within a library/process can be critical in situations with arbitrary messages (using the self-describing message technique) where code processing a proto encoded as arbitrary bytes would then do some kind of fully-qualified-name lookup to resolve the type.

Introducing a new option to support this use case in Swift when no other language supports or needs it would be strange; it would be akin to wanting a way to use two Java classes with the same package name and class name in the same process. It just doesn't come up that often, because well-formed protos are named in such a way that these collisions don't occur.

The reason many of the test suites don't unconditionally bring in all the test protos is because not all language implementations support the full set of features in the union of all those protos. For example, the Java implementation doesn't need the C++ arena protos, and other languages don't need to deal with the Java lite runtime.

And while users could manage their own typealiases, as the number of messages and types in a project becomes large, I would *much* rather specify a single option that creates them for me automatically over maintaining my own list manually. The need for such a feature is driven by Swift's unfortunate current state of not having very good support for syntactically-driven namespaces. I don't find it compelling to want an option to distinguish between multiple generated types for messages with the same fully-qualified name—a situation that has only occurred in a specific testing scenario and can be easily worked around—but then to find an option that automatically creates convenience typealiases unnecessary, requiring users to curate a potentially long list of aliases.

Given that the prefix and the proto are inextricably linked by virtue of being in the same .proto file, I probably wouldn't have any objections to choosing the replacement option vs. adding a typealias.